### PR TITLE
Add Azure Resource Logs translator

### PR DESCRIPTION
**Pull request by MikeGoldsmith**
**Description**:

This PR adds a new Azure Resource Logs translator that can be used to convert Azure events into resource logs events using Semantic conventions defined here [here](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/azure/events.md).

This supersedes following draft PR by updating it to have the latest semconv:
- #32486

*Note*: A follow-up PR will update the Azure receiver to use the new translator.

**Testing**:

Includes unit tests to verify expected behaviours and data structures.

**Documentation**:

---

cc @markrendle @lmolkova @TylerHelmuth

### CODE REVIEW

**Comment by MikeGoldsmith** [comment]
I'm a little unsure why both the changelog is failing:
image

and why the receiver is failing with a missing go.sum entry
image

Any help would be appreciated 😄

**Comment by cparkins** [comment]
@MikeGoldsmith I would remove the '[user]' from your change log entry and see if it fixes the issue since it says it's the default value anyway. I'll also look into the go mod issue, although I always struggle with those too!

**Comment by atoulme** [comment]
single quotes around your array makes it a string `'[user]'`

**Comment by cparkins** [comment]
@MikeGoldsmith I submitted a PR to your fork to try and address some of the build issues.

**Comment by MikeGoldsmith** [comment]
Thank you @cparkins, I'll take a look shortly.

**Comment by MikeGoldsmith** [comment]
@cparkins @atoulme looks like I still have some weird issues, any more advice?

I can work around the CLA issue by squashing the extra commit.

File: cmd/otelcontribcol/go.mod
Comment (codeboten) [inline cmd/otelcontribcol/go.mod:7]:
```suggestion
toolchain go1.22.6
```

File: cmd/oteltestbedcol/go.mod
Comment (codeboten) [inline cmd/oteltestbedcol/go.mod:7]:
```suggestion
toolchain go1.22.6
```

**Comment by MikeGoldsmith** [comment]
Okay, I think I've fixed it all up now. That was tough.

Please could someone update the labels to remove the extras - I think all we need are:
- cmd/otelcontribcol
- receiver/azureeventhub

There will be a new pkg/translator/azurelogs after this is merged.

**Comment by crobert-1** [comment]
From failing check:
```
Dropdowns in issue templates are out of date, please run "make generate-gh-issue-templates" and commit the changes in this PR.
```

**Comment by MikeGoldsmith** [comment]
@cparkins @atoulme @crobert-1 Looks like we've got green tests 🎉 

I'd appreciate it if you could review the actual code now 😄

File: pkg/translator/azurelogs/metadata.yaml
Comment (crobert-1) [inline pkg/translator/azurelogs/metadata.yaml:3]:
```suggestion
    active: [atoulme, cparkins]
```
The `open-telemetry/collector-approvers` group is added automatically to the code owners file, it doesn't need to be included here.

File: .github/CODEOWNERS
Comment (crobert-1) [inline .github/CODEOWNERS:159]: The indentation looks off here, it should be lined up with other lines.

File: pkg/translator/azurelogs/normalize.go
Comment (crobert-1) [inline pkg/translator/azurelogs/normalize.go:1]: Is there an existing package that handles these kinds of conversions? It seems like pretty generic functionality, especially `toLower`, `toFloat`, and `toInt`.

**Review by crobert-1 — COMMENTED** [review]
Just did a quick scan review. If you have a hard time getting timely reviews it would probably be helpful to split this PR into smaller chunks if possible.

File: pkg/translator/azurelogs/normalize.go
Comment (MikeGoldsmith) [reply pkg/translator/azurelogs/normalize.go:1]: I've cleaned up the normalize code a little to use a switch instead of a map in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/34830/commits/f99d3a631bf456bb9daa312ec33ebb63efab4567

I don't know of any utilities that can coerce `any` into a given type - happy to take a suggestion if you have one.

**Comment by MikeGoldsmith** [comment]
Thanks for the feedback @crobert-1. Let me know if you have more.

I can break into two PRs if you think it would help:
1) Add new azurelogs translator, not used yet
2) Update azureeventsreceiver to use new translator

This follows the tracks with the two change log entries as one is adding a new component (translator) and the other is updating an existing component (receiver).

**Comment by crobert-1** [comment]
> I can break into two PRs if you think it would help:

Sure that would definitely be helpful. I just mention it in case you don't get reviews in a timely manner, it can be hard to review larger PRs like this. Smaller PRs often get through the process quicker 👍

**Comment by an-mmx** [comment]
Hey guys!

This change is a great work and our team is much interested in this!

We have try these changes in our custom otlp build. We have reading `FunctionAppLogs` and there is another challenge for us to support US datetime format. However, I'd like to highlight another thing. Even though we were able to parse the US datetime format, we've faced issue with `azure.properties` field format. Because of different OS/SKU, the `azure.properties` might have different format. Sometimes it is string, sometimes it has non-json quotation (you may found some details here: https://github.com/Azure/azure-functions-host/issues/7864). This leads to panic upon reading the `azure.properties`. It seems microsoft doesn't care about having the same format for neither `time` nor `azure.properties`. Could you clarify how are you going to handle these differences in formats?
To be clear, we are going to use this new transform with some addition. And we are ready to contribute or share our thoughts.
In our custom implementation, we agreed on having a configuration option for custom time format. It should specify logs category and associated custom time format, eg: `FunctionAppLogs: "01/02/2006 15:04:05"`. As for the `azure.properties`, we are going to add extra type check. And for the case with the `string` type we will ensure proper quotation and extra json parsing.
How do you feel about introducing these logic in this mr?

**Comment by MikeGoldsmith** [comment]
@an-mmx I think I'm ready to incorporate your suggestions into this PR now.

I've improved the type handling when parsing `properties` in [5dd53f1](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/34830/commits/5dd53f178ce7c49539b6853673cb3cb31b68c5e4). Let me know what you think.

Do you have an example how of you did the timestamp parsing with different formats?

**Comment by MikeGoldsmith** [comment]
@crobert-1 I've updated this PR to only add the new translator. Please can you remove the receiver/azureeventshub label please?

Also noting that I'm seeing prometheus-compliance-test failures which is unrelated to my changes. I believe the following issue is tracking that being fixed:
- https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/35119

**Comment by MikeGoldsmith** [comment]
@crobert-1 how do you feel about the PR as it is now? Is there anything else you'd want to see before we consider this ready?

Regarding the feedback from @an-mmx, I've improved type handling to prevent panic's on unknown types and we could add better timestamp parsing in a follow up PR. I'm happy to handle that.

The plan is to update the AzureEventsReceiver to use this new translator in a separate PR.

Also, anymore reviews from @open-telemetry/collector-contrib-approvers would be appreciated.

**Review by atoulme — APPROVED** [review]

**Review by codeboten — APPROVED** [review]

**Review by TylerHelmuth — APPROVED** [review]

**Review by codeboten — APPROVED** [review]

**Comment by an-mmx** [comment]
Hey @MikeGoldsmith ,

Sorry for the late reply. It seems i've missed some updates.
I've seen your changes for `properties` type check and this probably a good solution. The `properties` of `FunctionAppLogs` is now stored as string attribute.

I believe it would be a good in next step to introduce the possibility to enable corrections to JSON string so that it would be possible to apply semantic conventions translation. We do it unconditionally in our implementation (you can check it in [this commit](https://github.com/an-mmx/opentelemetry-collector-contrib/commit/1108f7998b0f70fb010969e5a85ee9499aca94cd)). I agree, It's not the best approach. It worth to add some extra logic to make it optional and keep the `properties` in attributes when failed to parse JSON. Nevertheless it allows us to apply translations. I'd like to hear your thoughts on this.

There is also [another commit](https://github.com/an-mmx/opentelemetry-collector-contrib/commit/13d8fec3b25cfe842f1ffde0df97103803167af0) with the time convertion approach. I would like to create a PR with these changes for the both `azure` and `azurelogs` translators, after #35357 is being merged. Of curse, if no objections?

**Pull request merged**
